### ISSUE TITLE
llvm-general doesn't build with cabal>= 1.21

### ISSUE THREAD

**Issue by cartazio** [bscarlet/llvm-general#122]
1.22 (stable) is due for a release in the next week or three, i'll talk with them to understand what might needs be changed

```
1 of 1] Compiling Main             ( dist/setup/setup.hs, dist/setup/Main.o )

dist/setup/setup.hs:129:7:
    Couldn't match type ‘IO b0’ with ‘TestFlags -> IO ()’
    Expected type: IO () -> IO b0 -> TestFlags -> IO ()
      Actual type: IO () -> IO b0 -> IO b0
    In a stmt of a 'do' block:
      addLLVMToLdLibraryPath (configFlags localBuildInfo)
    In the expression:
      do { addLLVMToLdLibraryPath (configFlags localBuildInfo);
           testHook
             simpleUserHooks
             packageDescription
             localBuildInfo
             userHooks
             testFlags }
    In the ‘testHook’ field of a record

dist/setup/setup.hs:129:43:
    Couldn't match expected type ‘LocalBuildInfo’
                with actual type ‘PackageDescription’
    In the first argument of ‘configFlags’, namely ‘localBuildInfo’
    In the first argument of ‘addLLVMToLdLibraryPath’, namely
      ‘(configFlags localBuildInfo)’

dist/setup/setup.hs:130:7:
    Couldn't match expected type ‘IO b0’
                with actual type ‘TestFlags -> IO ()’
    Probable cause: ‘testHook’ is applied to too few arguments
    In a stmt of a 'do' block:
      testHook
        simpleUserHooks
        packageDescription
        localBuildInfo
        userHooks
        testFlags
    In the expression:
      do { addLLVMToLdLibraryPath (configFlags localBuildInfo);
           testHook
             simpleUserHooks
             packageDescription
             localBuildInfo
             userHooks
             testFlags }
```

**Comment by erikd** [comment]
@ttuegel There may be an easier way than TH.

* In `setup.hs`:

    import Distribution.Simple.Utils
    import Text.Printf
    main = do
        writeFile "CabalVersion.h" ("#define CabalVersion " ++ 
                                                        printf "%2%02d%02d%02d" cabalVesion)
        system "runhaskell RealSetup.hs"

* In `RealSetup.hs`:

    #include "CabalVersion.h"

    #if CABAL_VERSION < 120000

    #else

    #endif

**Comment by ttuegel** [comment]
That should do it, but don't forget to pass on the command-line arguments!

**Comment by cartazio** [comment]
the CLI args detail makes me think the TH approach is easier. i'll try to find some time to cook up a PR later this week

**Comment by ttuegel** [comment]
```
getArgs >>= rawSystem "runhaskell" . ("RealSetup.hs" :)
```
or something.

**Comment by erikd** [comment]
@cartazio TH is not going to work on platforms where TH isn't working, nor is it going to work when cross-compiling (at least until TH and cross-compiling is fixed).

**Comment by cartazio** [comment]
fair enough

**Comment by bscarlet** [comment]
I haven't had time to build a trial version of my build environment with the new cabal version, so i don't yet have much insight into the change. Would a simple typeclass-and-instances-for-each-version approach (like the one in there already) work?

**Comment by bscarlet** [comment]
I haven't had time to build a trial version of my build environment with the new cabal version, so I don't yet have much insight into the change. Would a simple typeclass-and-instances-for-each-version approach (like the one in there already) work?

**Comment by bscarlet** [comment]
should be fixed on llvm-3.3, llvm-3.4, and master
pushed to hackage as llvm-general-3.3.13.3 and llvm-general-3.4.4.3

**Comment by cartazio** [comment]
oh, that a cute hack with type classes! I've not seen that approach before, clever, and elegant event!
